File: frontend/src/mocks/fixtures/transcript-coding.ts
import type {
  TranscriptCodeValue,
  TranscriptEvidenceValue,
  TranscriptHighlightValue,
  TranscriptReaderBlockValue,
} from "@/components/research/transcript-coding-types";
import type {
  RecordCode,
  TranscriptAnchor,
  TranscriptCodingWorkspace,
} from "@/api/types";

export const transcriptCodingCodes: TranscriptCodeValue[] = [
  {
    id: "code-navigation-terminology",
    name: "Navigation terminology",
    description: "Labels, grouping, and navigation language.",
  },
  {
    id: "code-workflow-confidence",
    name: "Workflow confidence",
    description: "Confidence while completing a workflow.",
  },
  {
    id: "code-information-architecture",
    name: "Information architecture",
    description: "Labels, grouping, and navigation.",
  },
];

export const transcriptCodingEvidence: TranscriptEvidenceValue[] = [
  {
    id: "evidence-navigation-primary",
    excerpt: "I expected uploaded transcripts to be under Documents, but the menu called it Source Material.",
    location: "12:45–13:02",
    speaker: "Jordan Moore",
  },
  {
    id: "evidence-navigation-secondary",
    excerpt: "I kept looking under Documents before I noticed that Source Material meant the raw transcript.",
    location: "14:08–14:22",
    speaker: "Jordan Moore",
  },
  {
    id: "evidence-confidence-primary",
    excerpt: "I paused because I was not sure whether Source Material meant the raw transcript or something already analyzed.",
    location: "13:16–13:38",
    speaker: "Jordan Moore",
  },
];

export const transcriptReaderBlocks: TranscriptReaderBlockValue[] = [
  {
    id: "reader-question-one",
    excerpt: "Where would you expect to find uploaded research after a session?",
    location: "12:31–12:44",
    speaker: "Maya Chen",
    state: "default",
  },
  {
    ...transcriptCodingEvidence[0],
    id: "reader-navigation",
    state: "selection-active",
  },
  {
    id: "reader-question-two",
    excerpt: "What made that label confusing in the moment?",
    location: "13:03–13:15",
    speaker: "Maya Chen",
    state: "default",
  },
  {
    ...transcriptCodingEvidence[2],
    id: "reader-confidence",
    state: "default",
  },
  {
    id: "reader-question-three",
    excerpt: "What label would have helped you continue with confidence?",
    location: "13:39–13:52",
    speaker: "Maya Chen",
    state: "default",
  },
];

export const importedTemplatedTranscriptBlocks: TranscriptReaderBlockValue[] = [
  {
    id: "imported-s003-1",
    excerpt: "Please send a secure message about your parent's medication.",
    location: "0:00",
    speaker: "Maya Chen",
    state: "default",
  },
  {
    id: "imported-s003-2",
    excerpt: "I am looking for Messages. I see it, but I want to confirm I am in my mother's account.",
    location: "3:06",
    speaker: "Tanya",
    state: "default",
  },
  {
    id: "imported-s003-3",
    excerpt: "What tells you which account is active?",
    location: "6:13",
    speaker: "Maya Chen",
    state: "default",
  },
  {
    id: "imported-s005-role-label",
    excerpt: "The recurring issue is that teams treat each Record as unique, but users experience one system.",
    location: "0:00",
    speaker: "Jordan - UX",
    state: "default",
  },
];

export const transcriptCodingSuggestions = [
  {
    id: "suggestion-navigation",
    codeName: "Navigation terminology",
    confidence: 0.88,
    description: "Labels for uploaded research did not match participant expectations.",
    evidence: transcriptCodingEvidence.slice(0, 2),
    provenance: "AI generated • Jordan Moore",
  },
  {
    id: "suggestion-confidence",
    codeName: "Workflow confidence",
    confidence: 0.88,
    description: "The participant paused because the destination label did not communicate what happened next.",
    evidence: transcriptCodingEvidence.slice(2, 3),
    provenance: "AI generated • Jordan Moore",
  },
] as const;

export const transcriptCodingHighlights: TranscriptHighlightValue[] = [
  {
    id: "highlight-navigation",
    codes: [transcriptCodingCodes[0]],
    evidence: transcriptCodingEvidence[0],
    provenance: "Researcher accepted",
    status: "accepted",
  },
  {
    id: "highlight-confidence",
    codes: [transcriptCodingCodes[1]],
    evidence: transcriptCodingEvidence[2],
    provenance: "Researcher accepted",
    status: "accepted",
  },
  {
    id: "highlight-uncoded",
    codes: [],
    evidence: transcriptCodingEvidence[1],
    provenance: "Researcher highlighted",
    status: "uncoded",
  },
];

const fixtureTimestamp = "2026-07-20T12:00:00Z";
const fixtureChecksum = "sha256:transcript-coding-fixture";

function apiCode(code: TranscriptCodeValue): RecordCode {
  return {
    id: code.id,
    record_id: "record-1",
    name: code.name,
    description: code.description ?? null,
    status: "active",
    created_at: fixtureTimestamp,
    updated_at: fixtureTimestamp,
  };
}

const codingBlocks = transcriptReaderBlocks.reduce<Array<TranscriptCodingWorkspace["transcript"]["blocks"][number]>>(
  (blocks, block) => {
    const start = blocks.length === 0 ? 0 : blocks[blocks.length - 1].end_char + 1;
    blocks.push({
      id: block.id,
      chunk_id: block.id,
      speaker: block.speaker ?? "Speaker",
      location: block.location,
      text: block.excerpt,
      start_char: start,
      end_char: start + block.excerpt.length,
    });
    return blocks;
  },
  [],
);

function anchorForEvidence(evidence: TranscriptEvidenceValue): TranscriptAnchor {
  const block = codingBlocks.find((candidate) =>
    candidate.id === evidence.blockId || candidate.text === evidence.excerpt,
  );
  return {
    document_id: "checkout-transcript",
    chunk_id: block?.chunk_id ?? null,
    block_id: block?.id ?? evidence.blockId ?? null,
    start_char: block?.start_char ?? 0,
    end_char: (block?.start_char ?? 0) + evidence.excerpt.length,
    excerpt_snapshot: evidence.excerpt,
    speaker: evidence.speaker ?? null,
    location: evidence.location,
    start_ms: null,
    end_ms: null,
    content_checksum: fixtureChecksum,
  };
}

export const transcriptCodingApiFixture: TranscriptCodingWorkspace = {
  project_id: "alpha-project",
  session_id: "mobile-checkout-test",
  document_id: "checkout-transcript",
  record: { id: "record-1", name: "Record 1" },
  transcript: {
    content_checksum: fixtureChecksum,
    blocks: codingBlocks,
  },
  suggestion_run: {
    id: "coding-run-1",
    status: "complete",
    error_detail: null,
  },
  codes: transcriptCodingCodes.map(apiCode),
  highlights: transcriptCodingHighlights.map((highlight) => ({
    id: highlight.id,
    project_id: "alpha-project",
    session_id: "mobile-checkout-test",
    document_id: "checkout-transcript",
    anchor: anchorForEvidence(highlight.evidence),
    origin: highlight.provenance?.includes("AI") ? "ai-suggestion" : "researcher",
    codes: highlight.codes.map(apiCode),
    created_at: fixtureTimestamp,
    updated_at: fixtureTimestamp,
    deleted_at: null,
  })),
  suggestions: transcriptCodingSuggestions.map((suggestion, suggestionIndex) => ({
    id: suggestion.id,
    run_id: "coding-run-1",
    record_id: "record-1",
    proposed_name: suggestion.codeName,
    proposed_description: suggestion.description,
    confidence: suggestion.confidence,
    status: "awaiting-review",
    was_edited: false,
    accepted_code_id: null,
    reviewed_at: null,
    evidence: suggestion.evidence.map((evidence, evidenceIndex) => ({
      id: evidence.id,
      anchor: anchorForEvidence(evidence),
      display_order: suggestionIndex * 10 + evidenceIndex,
    })),
  })),
};

File: backend/app/schemas/transcript_coding.py
from datetime import datetime
from typing import Literal

from app.schemas.research_session import SessionReference
from pydantic import BaseModel, Field, model_validator


class TranscriptCodingBlock(BaseModel):
    id: str
    chunk_id: str | None
    speaker: str
    location: str
    text: str
    start_char: int
    end_char: int


class TranscriptAnchor(BaseModel):
    document_id: str
    chunk_id: str | None
    block_id: str | None
    start_char: int
    end_char: int
    excerpt_snapshot: str
    speaker: str | None
    location: str | None
    start_ms: int | None
    end_ms: int | None
    content_checksum: str


class TranscriptAnchorCreate(BaseModel):
    chunk_id: str | None = None
    block_id: str | None = None
    start_char: int
    end_char: int
    excerpt_snapshot: str
    speaker: str | None = None
    location: str | None = None
    start_ms: int | None = Field(default=None, ge=0)
    end_ms: int | None = Field(default=None, ge=0)
    content_checksum: str


class RecordCodeRead(BaseModel):
    id: str
    record_id: str
    name: str
    description: str | None
    status: Literal["active", "archived"]
    created_at: datetime
    updated_at: datetime


class RecordCodeSupportingHighlightRead(BaseModel):
    id: str
    project_id: str
    project_name: str
    session_id: str
    session_title: str
    excerpt: str
    speaker: str | None
    location: str | None


class RecordCodeEvidenceGroupRead(BaseModel):
    session_id: str
    session_title: str
    highlights: list[RecordCodeSupportingHighlightRead]


class RecordAcceptedCodeRead(BaseModel):
    id: str
    record_id: str
    name: str
    description: str | None
    accepted_highlight_count: int
    session_count: int
    latest_evidence_at: datetime
    evidence_groups: list[RecordCodeEvidenceGroupRead]


class RecordTranscriptCodesRead(BaseModel):
    record_id: str
    accepted_code_count: int
    session_count: int
    codes: list[RecordAcceptedCodeRead]


class RecordCodeCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=1200)


class RecordCodeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=1200)


class NewRecordCode(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=1200)


class TranscriptHighlightRead(BaseModel):
    id: str
    project_id: str
    session_id: str
    document_id: str
    anchor: TranscriptAnchor
    origin: Literal["researcher", "ai-suggestion"]
    codes: list[RecordCodeRead]
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class TranscriptHighlightCreate(BaseModel):
    anchor: TranscriptAnchorCreate
    code_ids: list[str] = Field(default_factory=list)
    new_code: NewRecordCode | None = None


class TranscriptHighlightUpdate(BaseModel):
    code_ids: list[str]


class HighlightCodesUpdate(BaseModel):
    code_ids: list[str] = Field(min_length=1)


class TranscriptCodeSuggestionEvidenceRead(BaseModel):
    id: str
    anchor: TranscriptAnchor
    display_order: int


class TranscriptCodeSuggestionRead(BaseModel):
    id: str
    run_id: str
    record_id: str
    proposed_name: str
    proposed_description: str | None
    confidence: float | None
    status: Literal["awaiting-review", "accepted", "rejected"]
    was_edited: bool
    accepted_code_id: str | None
    reviewed_at: datetime | None
    evidence: list[TranscriptCodeSuggestionEvidenceRead]


class TranscriptCodeSuggestionUpdate(BaseModel):
    proposed_name: str | None = Field(default=None, min_length=1, max_length=120)
    proposed_description: str | None = Field(default=None, max_length=1200)

    @model_validator(mode="after")
    def require_change(self):
        if "proposed_name" not in self.model_fields_set and "proposed_description" not in self.model_fields_set:
            raise ValueError("Provide a proposed name or description.")
        return self


class SuggestionRunRead(BaseModel):
    id: str | None
    status: Literal["idle", "queued", "processing", "complete", "failed"]
    error_detail: str | None


class TranscriptCodingTranscriptRead(BaseModel):
    content_checksum: str
    blocks: list[TranscriptCodingBlock]


class TranscriptCodingWorkspaceRead(BaseModel):
    project_id: str
    session_id: str
    document_id: str
    record: SessionReference | None
    transcript: TranscriptCodingTranscriptRead
    suggestion_run: SuggestionRunRead
    codes: list[RecordCodeRead]
    highlights: list[TranscriptHighlightRead]
    suggestions: list[TranscriptCodeSuggestionRead]
